### PR TITLE
Cloud spawns happen on regular intervals, and conform to the compound densities of a patch. (closes 1560)

### LINKED ISSUE #1560: Make cloud spawns less random and more time based
Instead of the totally chance dependent cloud spawning I think we should move to a timer based approach that would guarantee clouds spawn in proportion with their relative densities and some (configurable) max rate.

A bit similar to: https://github.com/Revolutionary-Games/Thrive/issues/1551

### PR DESCRIPTION
**Pull request by Gwen-Benzschawel**
This replaces the old system for one that's like a bag. Compounds are added to the bag and shuffled. When the spawn er needs a cloud, it draws one from the bag. When the bag is empty, the bag is refilled and shuffled. This ensures that you can get a steady rate of each compound, but the order of spawn is still random. 

The compound bag is set up so that it will empty the bag at a predefined constant rate that can be changed for game-play balance.

closes #1560

### CODE REVIEW

File: src/microbe_stage/MicrobeStage.cs [hunk 1]
```diff
@@ -252,7 +252,7 @@ public void SetupStage()
         if (IsLoadedFromSave)
         {
             HUD.OnEnterStageTransition(false);
-            UpdatePatchSettings(true);
+            UpdatePatchSettings(true, Player.Transform.origin);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:255]: Why did you change this to require the player position? The player position is already passed in on the spawn cycle, it shouldn't require setup here.

File: simulation_parameters/Constants.cs [hunk 2]
```diff
@@ -46,6 +46,11 @@ public static class Constants
     // Should be the same as its counterpart in shaders/CompoundCloudPlane.shader
     public const float CLOUD_MAX_INTENSITY_SHOWN = 1000;
 
+    // length of cloud spawn cycle in seconds.
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:49]: Comment should start with a capital letter.

File: src/microbe_stage/PatchManager.cs [hunk 3]
```diff
@@ -105,23 +103,39 @@ private void HandleChunkSpawns(BiomeConditions biome)
         }
     }
 
-    private void HandleCloudSpawns(BiomeConditions biome)
+    private void HandleCloudSpawns(BiomeConditions biome, Vector3 playerPosition)
     {
         GD.Print("Number of clouds in this patch = ", biome.Compounds.Count);
 
-        foreach (var entry in biome.Compounds)
+        spawnSystem.ClearBiomeCompounds();
```
Comment (hhyyrylainen) [inline src/microbe_stage/PatchManager.cs:110]: This shouldn't always happen here, this should only be triggered when the spawners are destroyed on patch change.

File: src/microbe_stage/Spawners.cs [hunk 4]
```diff
@@ -342,24 +341,27 @@ public override IEnumerable<ISpawned> Spawn(Node worldNode, Vector3 location)
 /// <summary>
 ///   Spawns compound clouds of a certain type
 /// </summary>
-public class CompoundCloudSpawner : Spawner
+public class CloudSpawner : Spawner
```
Comment (hhyyrylainen) [inline src/microbe_stage/Spawners.cs:344]: Couldn't the design work so that the CompoundCloudSpawner class is left as is, with the SpawnSystem handling the selecting which clouds are allowed to spawn?

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Overall this makes the spawn system *much* more complicated as there's now in fact two separate systems in there: the clouds and other stuff. I'd prefer to go like all the way at once (also make cells and chunks spawn using a timer with weighted distribution), or retrofit the new cloud logic as much as possible to the existing system.

File: src/microbe_stage/Spawners.cs (hunk 4 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/Spawners.cs:344]: The old system had multipleCompoundCloudSpawners, one per compound. The new system has only one CloudSpawner, where the compounds are selected with the cloudBag. I figured this justified the name change.

File: src/microbe_stage/PatchManager.cs (hunk 3 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/PatchManager.cs:110]: I've heard the eventual goal is to have compound amounts and densities change over time. This would mean clearing the system here. I guess you are right though, that's currently not a feature, so it makes no sense now.

File: simulation_parameters/Constants.cs (hunk 2 above)
Comment (Gwen-Benzschawel) [reply simulation_parameters/Constants.cs]: Fixed.

File: src/microbe_stage/Spawners.cs (hunk 4 above)
Comment (hhyyrylainen) [reply src/microbe_stage/Spawners.cs:344]: I did gather that from the changes. However, I think in order to decouple the spawn setup, from the actual spawning logic, you'll need to reintroduce back this intermediate CompoundCloudSpawner per type. Because as I said the spawn setup should *not* care at all what the player position is. The spawn system can have internally used a CloudSpawner that keeps track of extra state that is constructed from the registered CompoundCloudSpawner instances.

I'm being pretty hard on these changes, because the spawn system is a huge problem, and making it extra complicated with special logic for one type of spawning, might mean that it will be even more difficult to get someone to work on the spawn system again. Even with the current logic, which I think is pretty minimal for the random spawning it currently does, it's been years that I've tried to nudge people into working on the system. That's why I'd like to see a unified spawner design to avoid unnecessary complexity that will make future changes harder.

Though this issue mentions only clouds: https://github.com/Revolutionary-Games/Thrive/issues/1560 I think people would welcome having all the spawners moved over to a time based spawn algorithm that would still take the different densities into account.

File: src/microbe_stage/MicrobeStage.cs (hunk 1 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/MicrobeStage.cs:255]: When testing my changes, it felt weird when you change patches and the screen is empty. So I decided to quickly added some free cloud spawns on patch change. In hindsight, yeah this isn't a good way to implement that.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs:255]: I think it is a good change to make some stuff spawn on patch change. I *think* in the cloud setup, there's already a parameter or something that detects when the patch changes (as it should clear the old spawners). At least the code that calls the patch setup should know if a patch change is happening, and could pass that information to either here or the spawn system, which might be a bit nicer way to do it (basically reset the spawn intervals to cause something to spawn immediately).

**Comment by Gwen-Benzschawel** [comment]
I think it wouldn't be too hard to just switch over every Spawner to this system. I could have a classes of Spawns, and the bag system can just pop and handle the Spawn. This would remove the Spawners themselves and simplify alot.

**Comment by hhyyrylainen** [comment]
> This would remove the Spawners themselves and simplify alot.

But you still would need some kind of tracking for what things want to spawn, right? I think quite a bit of the SpawnSystem state is not saved, instead the patch settings are re-applied when a save is loaded. That could be changed, but then the patch settings object would be much more tightly tied to the spawn system as it needs to look at its internal state to determine which spawns to add, remove and adjust.

File: src/microbe_stage/Spawners.cs (hunk 4 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/Spawners.cs:344]: I understand why you are being hard on this, something like a spawn system should be easy to understand and change. I will rework it when I have free time.

Comment (hhyyrylainen) [reply src/microbe_stage/Spawners.cs:344]: I look forward to that. It would be great to finally get the spawn system into better shape.
While pondering this a bit more, I think it could in fact simplify the design if the spawners are owned by the spawn system (and stored in saves, though getting delegates to save properly is not entirely trivial), and then the patch conditions could read the existing data, and communicate through method calls the necessary updates (as the spawn system probably needs to react to the changes, so it isn't a good idea to let the patch conditions applier to directly modify the spawner properties). But anyway, if you don't want to dig into the saving part yet, you can keep the current intermediate spawner objects, which as I said, make the behaviour regarding saving easier to *not* break.

**Comment by buckly90** [comment]
I have tested this PR in-game and am pleased with how some compounds immediately spawn upon migration.

However; Compound spawn rates seem to have diminished in occurrence overall making all patches equally devoid of sufficient compound clouds. It would also be nice if the system included chunk spawning as well, but maybe you just didn't get to that yet.

I feel like this method could work assuming that the interval is easily tunable. But it is difficult to fully gauge how much more effective it is than our current system with the current scarcity of resulting resources.

**Comment by Gwen-Benzschawel** [comment]
Update: I have decided that, since nobody seems to like the old spawn code, I'm just going to rewrite it. On my local branch I have already removed the old system and added chunk spawning to the new system. I also found a bug with this current PR, where cloud spawning rate would decrease over time. I fixed it already in my re-write. Hopefully I will find the time to add microbe spawns and then you guys will have a shiny new spawn system. :)

**Comment by Nunez2196** [comment]
thank god

**Comment by Gwen-Benzschawel** [comment]
And that's the spawn system rewrite. Its just as long as the original, but it spawns clouds, chunks, and microbes in a time based manner. For most of the frames, the SpawnSystem is just checking a timer. Once that timer goes off, it picks a spawn out of the shuffled SpawnItemsBag, and then spawns it. The SpawnItem could be a cloud, chunk or microbe. All of the data needed to spawn the item are stored in the SpawnItem. The spawn system will only ever spawn one item per frame at maximum. No more queued spawn. When the bag is empty, the SpawnSystem needs to fill it. This is the only part that could cause lag, because it needs to both generate a bunch of SpawnItems, and also shuffle them in the bag. By keeping the bag small and using an efficient shuffling algorithm, I don't think that will be a problem.

Overall, I think this system will reduce lag and make resource spawns feel more consistent. I also added a function to spawn stuff in if you are changing patches, so new patches do not feel like empty voids. Hopefully this will be a better system that is easier to maintain.

**Comment by Gwen-Benzschawel** [comment]
I was looking at the other span related issues. This new system definitely makes iron chunks spawn. Honestly in some patches there is now too much iron. This system should also allow for a more robust encounter based design (#2055) by making the SpawnItem classes more complex. Like maybe a large iron chunk spawn also spawns in microbes with several Rusticyanin if a species exists, for example.

File: src/microbe_stage/SpawnSystem.cs [hunk 5]
```diff
@@ -350,10 +303,10 @@ private int DespawnEntities(Vector3 playerPosition)
             }
 
             var entityPosition = ((Spatial)entity).Translation;
-            var squaredDistance = (playerPosition - entityPosition).LengthSquared();
+            var distance = (playerPosition - entityPosition).Length();
```
Comment (hhyyrylainen) [inline src/microbe_stage/SpawnSystem.cs:306]: LengthSquared should absolutely be kept here as Length is *much* more expensive to compute as it requires as square root computation.

**Comment by hhyyrylainen** [comment]
Good work on this.
I'll ask for some playtesting on this, before that's complete I only quickly took a look at the code (and besides that one comment I left, I have a few others):
- It seems you got rid of the queued spawns. Why is that? When microbe colonies spawn they cause a lag spike, so that's why the queued spawn code is needed.
- I'd like a bit more of the smartness to be in the patch manager, the spawn system shouldn't really care about microbe species and such, I'd like the system to be generalizable to other situations as well. So I think it should be less tightly linked to the exact classes. Instead the intermediate spawner type should be the only thing the spawn system cares about, with the PatchManager creating the spawners based on the actual object types.
- I think some of the initialization logic is a bit wonky: instead of immediately spawning stuff (and taking in the player position), I think the spawn system should set an internal flag that patch has changed and a bigger spawn should happen on the next frame. And also the moving of the init logic around out of the non-save loaded guard. Why's that? I'd prefer if the spawn system actually saved its state so that another initialization is not needed after load from save. As to me that seems to imply things may not be saved as comprehensively as they could. It might be even worth considering that the concrete spawners could be saved, as I wrote a system that allows saving delegates (used for the editor actions).

File: src/microbe_stage/SpawnSystem.cs (hunk 5 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/SpawnSystem.cs]: Oh I remember what I was thinking. I changed spawning so it would use distance without ever needing to call Length or a sqrt function. But then I didn't think about despawning when I changed it... 

I just fixed that silly mistake.

**Comment by Gwen-Benzschawel** [comment]
Thanks! Yeah I hope you guys find this system useful. A quick note on playtesting, I did not change the cloud/chunk densities from the old system, and I also reused the formula that calculates densities from microbe population. These can be easily changed, but we still need to dial this system in.

>  It seems you got rid of the queued spawns. Why is that? When microbe colonies spawn they cause a lag spike, so that's why 
>  the queued spawn code is needed.

Yeah... When I removed them I had just got clouds and chunks working. I guess I thought that system was to ensure that all these random spawners don't all decide to spawn at once. At that time I didn't really look at the microbe spawner. Hmm... Well now that I know it's purpose was for microbe colony lag, yeah we really need to either add it back in or change how microbe colonies are spawned...

>  I'd like a bit more of the smartness to be in the patch manager, the spawn system shouldn't really care about microbe
>  species and such, I'd like the system to be generalizable to other situations as well. So I think it should be less tightly linked 
>  to the exact classes. Instead the intermediate spawner type should be the only thing the spawn system cares about, with the 
>  PatchManager creating the spawners based on the actual object types.

I think that can be done by moving the FillSpawnItemBag() function and the compound/chunk/species lists over to the patch manager. The PatchManager would then fill the spawn bag and give it to the SpawnSystem, which could then just ask for a new one once it runs out.

>  I think some of the initialization logic is a bit wonky: instead of immediately spawning stuff (and taking in the player 
>  position), I think the spawn system should set an internal flag that patch has changed and a bigger spawn should happen 
>  on the next frame.

Yeah, that would be a better implementation.

>  ...And also the moving of the init logic around out of the non-save loaded guard. Why's that? I'd prefer if the spawn system 
>  actually saved its state so that another initialization is not needed after load from save. As to me that seems to imply things 
>  may not be saved as comprehensively as they could. It might be even worth considering that the concrete spawners could 
>  be saved, as I wrote a system that allows saving delegates (used for the editor actions).

I'm not that familiar with Newtonsoft.Json. I tried to get saving the SpawnItemBag implemented. It sort of worked, I could save and load SpawnItems in the bag using a Json Converter, but I couldn't get the deseralized SpawnItems to actually spawn without crashing. At this point I had already spent a few hours on this problem before I just decided to remove the serialization and push it to you guys for feedback on the actual spawn system. Maybe when I have more time I could implement serialization, but for now saving and loading will reset the spawn bag, causing different items to spawn than if you had kept playing.

**Comment by hhyyrylainen** [comment]
> I'm not that familiar with Newtonsoft.Json. I tried to get saving the SpawnItemBag implemented. It sort of worked, I could save and load SpawnItems in the bag using a Json Converter, but I couldn't get the deseralized SpawnItems to actually spawn without crashing. At this point I had already spent a few hours on this problem before I just decided to remove the serialization and push it to you guys for feedback on the actual spawn system. Maybe when I have more time I could implement serialization, but for now saving and loading will reset the spawn bag, causing different items to spawn than if you had kept playing.

It is currently like that as well (but the re-initialization happens by PatchManager so the spawn system init doesn't need to run twice), but now that I've implemented that delegate saving. Basically to save stuff, you'll need a base class for spawn system state, then add that in the spawner, and the spawner then also has a delegate that is saved. The delegate would point to a concrete spawn implementation. I hope I made it clear enough in the MicrobeEditor how it works in there, it's actually pretty simple just requiring one attribute for the delegate target to allow it to work. But yeah, not a huge priority, I'd just like to at least see the duplicate initialization not needing to be done. The spawn system could be marked with the ISaveLoadable interface so it can receive a callback when loaded from save to reset some state.

> I think that can be done by moving the FillSpawnItemBag() function and the compound/chunk/species lists over to the patch manager. The PatchManager would then fill the spawn bag and give it to the SpawnSystem, which could then just ask for a new one once it runs out.

Referring back to the old design: I was thinking that you would have a SpawnerConfiguration class and instance that when passing to the spawn system. Then the spawn system could make whatever bag shuffling stuff it needs based on that. So it would only depend on properly created SpawnerConfigurations than the concrete microbe etc. classes.

> Thanks! Yeah I hope you guys find this system useful. A quick note on playtesting, I did not change the cloud/chunk densities from the old system, and I also reused the formula that calculates densities from microbe population. These can be easily changed, but we still need to dial this system in.

First gameplay testing feedback is that this system definitely feels a bit different. So once the basic structure is done, we'll probably want to do some spawn density tweaking to balance things.

File: src/microbe_stage/SpawnSystem.cs [hunk 6]
```diff
@@ -303,10 +303,10 @@ private void DespawnEntities(Vector3 playerPosition)
             }
 
             var entityPosition = ((Spatial)entity).Translation;
-            var distance = (playerPosition - entityPosition).Length();
+            var squaredDistance = (playerPosition - entityPosition).LengthSquared();
 
             // If the entity is too far away from the player, despawn it.
-            if (distance > spawned.DespawnRadius)
+            if (squaredDistance > spawned.DespawnRadius * spawned.DespawnRadius)
```
Comment (hhyyrylainen) [inline src/microbe_stage/SpawnSystem.cs:309]: In the old code `spawned.DespawnRadius` was already the squared distance so, that avoided also this one multiplication.

**Comment by hhyyrylainen** [comment]
Oh and I have one more general comment that I forgot earlier: I think the entity cap should be reintroduced, if you don't want to add per-type caps that's fine, but for the sake of people's computers we should have a global overall cap on the number of alive cells and existing chunks.

**Comment by buckly90** [comment]
I have just got done testing this PR, and while it's not perfect, I can say with certainty that it is an improvement.

Starting a new game, the world felt very scarce and empty, but after a while I began to encounter resources and it felt fine after that. I think the issue here is that everything was spawning to the side, and outside my view so I may not have noticed them.
This presumably isn't too big of a problem if that everything works correctly now! Just a matter of tweaking spawn rates to see if it helps.

I later discovered that sitting still allowed things to spawn around me indefinitely, ultimately resulting in a massive load of resources in one area. Reducing spawnrates while the player is still may prevent this.

I was very pleased to see that this fixes the missing iron problem in the caves.

**Comment by hhyyrylainen** [comment]
> Starting a new game, the world initially felt very scarce and empty, but after a while I began to encounter resources and it felt fine after that. I think the issue here is that everything was spawning to the side, and outside my view so I may not have noticed them.
This presumably isn't too big of a problem if everything works correctly now! Just a matter of tweaking spawn rates to see if it helps.

That reminds me that the old system had special logic to spawn things with a bigger chance to the player's front than the back. If that isn't in the new system, I think reimplementing that would be nice as it guarantees that the spawned entities that take up computing resources are more likely to be encountered by the player.

> I later discovered that sitting still allowed things to spawn around me indefinitely, ultimately resulting in a massive load of resources in one area. Reducing spawnrates while the player is still may prevent this.

There is actually a PR for this, but it never got finished: https://github.com/Revolutionary-Games/Thrive/pull/1739
I think the new system can be merged without that, but if going for perfection implementing that compound spawn reduce rate would make the system better.

**Comment by Gwen-Benzschawel** [comment]
>  That reminds me that the old system had special logic to spawn things with a bigger chance to the player's front than the back. 

This newer system has that, its the NewPatchSpawn() function, but because the spawn bag now has microbes and chunks there is a good chance that your starting spawns are 2 tiny iron chunks and a microbe that runs in the opposite direction.

**Comment by hhyyrylainen** [comment]
> This newer system has that, its the NewPatchSpawn() function

Ah, okay, good. I didn't read the code carefully enough yet to notice that.

**Comment by Gwen-Benzschawel** [comment]
I think my plan for this pr, is first to fix the concerns that @hhyyrylainen pointed out about the new system before I start tweaking it. 

I think it would be easy to fix the "spawn camping" issue. When a spawn happens, save the position. For the next spawn check to see if you moved much. If you haven't moved much, skip spawning chunks and compounds. Ideally I would like to find a way to keep spawning microbes so you can play as a species who doesn't move much if at all, surviving on microbes who wander to you.

Now, the "circle spawning" would be trickier, because you are moving just checking position in between spawns would not work. I think the only way to fix it would be to somehow check how dense an area is to decide if it should spawn something there. I think it could be somewhat tricky to figure out an elegant system for that, but should be doable.

**Comment by FuJa0815** [comment]
I've read through the code, but I think I need a bit more time to grasp it all.
The anti-camp protection is a must. The game ran in the background while I read the code and when I tabbed back, the result was... interesting :D
grafik

**Comment by Gwen-Benzschawel** [comment]
I changed it over to spawning things in clumps. I would like some playtesting on this to see if this is better than the old spawn circle, and what tweaks would need to be made.

It's not completely finished yet.
TODO:
Saving/Loading
"chunkiness" patch value that changes how big and far apart spawn clusters should be based on patch
wandering microbes that use the old system, so stationary player still see some microbe spawns

**Comment by hhyyrylainen** [comment]
This is out of date when compared to the master branch, meaning that this doesn't have the critical build script change I made for the new devcenter and that's why the build job is failing here.

**Comment by Gwen-Benzschawel** [comment]
The complete rewrite of the spawn system is done. It now uses an event similar to the node based one described here: https://forum.revolutionarygamesstudio.com/t/the-microbe-spawn-system/767/3

Wandering microbes are also supported. If you sit still in a void of nothing, a wandering microbe can appear on screen. Other than wandering microbes, nothing else will spawn if you sit still.

I changed the patch density values to work better with the new system. I have added a "chunkiness" and overall "spawnRateMultiplier" to make it easier to change the balance of a patch. Each spawnItem is probabilisticly independent from other spawnItems, meaning adding some sort of differential equations based method of changing compounds over time will just work without changing much of the spawn system.

Overall I think this spawn system makes the game feel alot better.

File: simulation_parameters/Constants.cs [hunk 7]
```diff
@@ -46,6 +46,29 @@ public static class Constants
     // Should be the same as its counterpart in shaders/CompoundCloudPlane.shader
     public const float CLOUD_MAX_INTENSITY_SHOWN = 1000;
 
+    // Length of wanderer microbe spawn cycle in seconds.
+    public const float WANDERER_SPAWN_RATE = 60f;
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:50]:
```suggestion
    public const float WANDERER_SPAWN_RATE = 60.0f;
```

File: simulation_parameters/Constants.cs [hunk 8]
```diff
@@ -46,6 +46,29 @@ public static class Constants
...
+    public const int SPAWN_GRID_HALFSIZE = SPAWN_GRID_SIZE / 2;
+
+    public const int SPAWN_EVENT_MIN = 10;
+    public const int SPAWN_EVENT_MAX = 15;
+
+    public const int SPAWN_GRID_WIDTH = 2;
+
+    public const int DESPAWN_ITEM_RADIUS = 170;
+    public const int DESPAWN_ITEM_RADIUS_SQR = DESPAWN_ITEM_RADIUS * DESPAWN_ITEM_RADIUS;
+    public const int SPAWN_WANDERER_RADIUS = (int)(0.95 * DESPAWN_ITEM_RADIUS);
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:64]: Would floating point precision be enough here instead of a double?

File: src/microbe_stage/AgentProjectile.cs [hunk 9]
```diff
@@ -18,6 +18,33 @@ public class AgentProjectile : RigidBody, ITimedLife
     [JsonProperty]
     private float? FadeTimeRemaining { get; set; }
 
+    public static AgentProjectile SpawnAgent(AgentProperties properties, float amount,
```
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:21]: If this is now included in this class this could be named just `Spawn`
this same comment probably applies to the other spawn methods as well.

File: src/microbe_stage/BiomeConditions.cs [hunk 10]
```diff
@@ -8,6 +8,8 @@
 public class BiomeConditions : ICloneable, ISaveLoadable
 {
     public float AverageTemperature;
+    public float SpawnChunkiness;
+    public float SpawnRateMultiplier;
```
Comment (hhyyrylainen) [inline src/microbe_stage/BiomeConditions.cs:12]: You should put the default values here, so the json could leave out values (if wanted in the future) where they are the defaults.

File: src/microbe_stage/ChunkSpawner.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        this.random = random;
+        chunkScene = LoadChunkScene();
+    }
+
+    public static PackedScene LoadChunkScene()
+    {
+        return GD.Load<PackedScene>("res://src/microbe_stage/FloatingChunk.tscn");
+    }
+
+    public static ISpawned SpawnChunk(ChunkConfiguration chunkType, Vector3 location, Node worldNode,
```
Comment (hhyyrylainen) [inline src/microbe_stage/ChunkSpawner.cs:25]: Why did you keep the separate class for this but not for the agentprojectile? I think for consistency one of the two approaches should be used, and not both.

File: src/microbe_stage/CompoundCloudSpawner.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+public class CompoundCloudSpawner
+{
+    private readonly CompoundCloudSystem clouds;
+
+    public CompoundCloudSpawner(CompoundCloudSystem clouds)
+    {
+        this.clouds = clouds ?? throw new ArgumentException("clouds is null");
+    }
+
+    public CompoundCloudSystem GetCloudSystem()
```
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundCloudSpawner.cs:16]: Getters should not be used. Properties with public get should be used instead. Though why is the clouds read out of this class? To me that seems a bit bad design, it should only be so that this class is given a CompoundCloudSystem object, but it isn't ever retrieved out of this class instance.

File: src/microbe_stage/MicrobeSpawner.cs [hunk 13]
```diff
@@ -179,85 +156,29 @@ public static PackedScene LoadMicrobeScene()
         return GD.Load<PackedScene>("res://src/microbe_stage/Microbe.tscn");
     }
 
-    public static FloatingChunk SpawnChunk(ChunkConfiguration chunkType,
-        Vector3 location, Node worldNode, PackedScene chunkScene,
-        CompoundCloudSystem cloudSystem, Random random)
+    public void SetCurrentGame(GameProperties currentGame)
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSpawner.cs:159]: Properties with public set should be used instead of setters.

File: src/microbe_stage/MicrobeSpawner.cs [hunk 14]
```diff
@@ -179,85 +156,29 @@ public static PackedScene LoadMicrobeScene()
...
+        this.currentGame = currentGame;
     }
 
-    /// <summary>
-    ///   Spawns an agent projectile
-    /// </summary>
-    public static AgentProjectile SpawnAgent(AgentProperties properties, float amount,
-        float lifetime, Vector3 location, Vector3 direction,
-        Node worldRoot, PackedScene agentScene, Node emitter)
+    public List<ISpawned> Spawn(Node worldNode, Vector3 location, MicrobeSpecies species, bool isWanderer)
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSpawner.cs:164]: This now creates the microbes all at once. The old system used enumerables to make sure that only the microbe instances that were to be spawned in that frame actually are spawned. I much prefer that earlier design as this could have problems regarding leaving undeleted microbes if the game is loaded / exited to menu while they are Microbe objects that have yet to be attached to the game scene.

File: src/microbe_stage/MicrobeStage.cs [hunk 15]
```diff
@@ -235,10 +235,10 @@ public void SetupStage()
         if (Settings.Instance == null)
             GD.PrintErr("Settings load problem");
 
+        spawner.Init(Clouds);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:238]: It looks like you added extra stuff for the spawn system to keep track of. However, you didn't remove the PatchManager also needing those resources. I think you should re-design this part so that the Spawn system does *not* know about these kinds of details, so this and the SetCurrentGame would be removed in favour of the PatchManager sending only the needed info to the SpawnSystem.

File: src/microbe_stage/Patch.cs [hunk 16]
```diff
@@ -147,7 +147,7 @@ public float GetTotalChunkCompoundAmount(Compound compound)
 
             if (chunk.Density > 0 && chunk.Compounds.ContainsKey(compound))
             {
-                result += chunk.Density * chunk.Compounds[compound].Amount;
+                result += chunk.Density;
```
Comment (hhyyrylainen) [inline src/microbe_stage/Patch.cs:150]: Why did you change this? This is for showing the compound amounts in the GUI (and maybe auto-evo uses this). But the chunk size as well as the density determines how much total compounds are available, right?

File: src/microbe_stage/MicrobeStage.cs [hunk 17]
```diff
@@ -319,9 +320,12 @@ public void SpawnPlayer()
         {
             // Random location on respawn
             var random = new Random();
-            Player.Translation = new Vector3(
+            Vector3 respawnPos = new Vector3(
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:323]: Having this intermediate variable is not needed, right?

File: src/microbe_stage/PatchManager.cs [hunk 18]
```diff
@@ -56,7 +58,7 @@ public void ApplyChangedPatchSettingsIfNeeded(Patch currentPatch, bool despawnAl
             // Despawn old entities
             spawnSystem.DespawnAll();
 
-            // And also all timed entities
+            // And also all timed entities and clear SpawnBags
```
Comment (hhyyrylainen) [inline src/microbe_stage/PatchManager.cs:61]: Isn't this comment change wrong?

File: src/microbe_stage/PatchManager.cs [hunk 19]
```diff
@@ -70,64 +72,81 @@ public void ApplyChangedPatchSettingsIfNeeded(Patch currentPatch, bool despawnAl
         // Update environment for process system
         processSystem.SetBiome(currentPatch.Biome);
 
-        // Apply spawn system settings
```
Comment (hhyyrylainen) [inline src/microbe_stage/PatchManager.cs:73]: I think this comment is still an useful label

File: src/microbe_stage/SpawnItem.cs [hunk 20]
```diff
@@ -0,0 +1,93 @@
+﻿using System.Collections.Generic;
+using Godot;
+
+// SpawnItems contains the data for one spawn.
```
Comment (hhyyrylainen) [inline src/microbe_stage/SpawnItem.cs:4]: xml comments need to be used on classes, functions, etc.

File: src/microbe_stage/editor/MicrobeEditorGUI.cs [hunk 21]
```diff
@@ -1874,11 +1874,9 @@ private float GetCompoundAmount(Patch patch, string compoundName)
             case "nitrogen":
                 return patch.Biome.Compounds[compound].Dissolved * 100;
             case "iron":
-                return patch.GetTotalChunkCompoundAmount(compound);
+                return patch.GetTotalChunkCompoundAmount(compound) * patch.Biome.SpawnRateMultiplier;
             default:
-                return patch.Biome.Compounds[compound].Density *
-                    patch.Biome.Compounds[compound].Amount + patch.GetTotalChunkCompoundAmount(
-                        compound);
+                return patch.Biome.Compounds[compound].Density * patch.Biome.SpawnRateMultiplier;
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorGUI.cs:1879]: You changed the logic here, why?

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Here's some code structure and style comments. I didn't yet have enough to read the patch system and spawn system class changes yet.

File: simulation_parameters/Constants.cs (hunk 8 above)
Comment (Gwen-Benzschawel) [reply simulation_parameters/Constants.cs:64]: I moved all of the spawn constants to one place and renamed this one. The old system had spawn radius stored as an int and do not know why. Yeah, I should have changed it.

File: src/microbe_stage/ChunkSpawner.cs (hunk 11 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/ChunkSpawner.cs:25]: Each SpawnItem points to their associated spawner so they can, well spawn. Also, a dying microbe needs to run the SpawnChunk code as well as the ChunkItem. That was my reason why I kept this but moved the agentprojectile.

File: src/microbe_stage/editor/MicrobeEditorGUI.cs (hunk 21 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/editor/MicrobeEditorGUI.cs:1879]: When playtesting, it felt like the percentages were "off", like if compounds have the same density you have equal chances of running into each cloud. But if they have different Amounts, then the percentages vary.

File: src/microbe_stage/ChunkSpawner.cs (hunk 11 above)
Comment (hhyyrylainen) [reply src/microbe_stage/ChunkSpawner.cs:25]: You can still reference static methods no matter which class they are in? Anyway I think only one of the approaches should be used because someone new reading the code may be confused by the same operation being in different places for different classes.

File: src/microbe_stage/editor/MicrobeEditorGUI.cs (hunk 21 above)
Comment (hhyyrylainen) [reply src/microbe_stage/editor/MicrobeEditorGUI.cs:1879]: But this is the *chunk* amount. Which heavily depends on how much resources each chunk has as the player can engulf it to get all of it's resources.

File: src/microbe_stage/PatchManager.cs (hunk 18 above)
Comment (Gwen-Benzschawel) [reply src/microbe_stage/PatchManager.cs:61]: opps. I think I did have it do that here at one point.

**Comment by Gwen-Benzschawel** [comment]
Ok I think I fixed all of the issues you have pointed out. As for the QueuedSpawn, right now I just added it into the current system. This means a spawn event (or in rare cases 2) fills up a Queue with SpawnItems, and when spawning if the SpawnItem has multiple spawns, an IEnumerator gets created and looped through over several frames.

**Comment by hhyyrylainen** [comment]
I think that was basically about how the current spawn system does the queued spawns over multiple frames.

**Comment by Gwen-Benzschawel** [comment]
Yeah, I basically just copied the existing system. I don't like how there is essentially 2 queue like things, but I can't think of a way to make it so each SpawnItem can only spawn one thing without adding more complexity and rewriting colony spawns. Oh well, it works and it's not that complicated.

**Comment by Untrustedlife** [comment]
> Yeah, I basically just copied the existing system. I don't like how there is essentially 2 queue like things, but I can't think of a way to make it so each SpawnItem can only spawn one thing without adding more complexity and rewriting colony spawns. Oh well, it works and it's not that complicated.

I have re-requested HH on this.

**Comment by hhyyrylainen** [comment]
Unless something was done about my concerns of leaking objects, that's a definite problem that needs addressing somehow (I suggested that design choice moving back to the enumerator style spawners).

**Comment by Untrustedlife** [comment]
> Unless something was done about my concerns of leaking objects, that's a definite problem that needs addressing somehow (I suggested that design choice moving back to the enumerator style spawners).

I thought they did, at least thats the jist I got from their comment. Sorry for bugging you hh, in that case nevertmind.

**Comment by hhyyrylainen** [comment]
> I thought they did, at least thats the jist I got from their comment. Sorry for bugging you hh, in that case nevertmind.

At least github doesn't show that any new commits were added after that comment 18 days ago, so I assume it wasn't addressed by any commit...

**Comment by 84634E1A607A** [comment]
I'm not sure if it's better to limit the spawning density. See what a mess I've got around me after leaving it untouched for a minute! (I'm sorry I commented in the wrong PR)

image

image

**Comment by hhyyrylainen** [comment]
Unless you specifically tested on this PR's code, then I don't think that problem is relevant to report here. Because this entirely changes the just chance based all things spawn randomly approach to a round robin timed bag of stuff to spawn approach.

**Comment by 84634E1A607A** [comment]
> Unless you specifically tested on this PR's code, then I don't think that problem is relevant to report here. Because this entirely changes the just chance based all things spawn randomly approach to a round robin timed bag of stuff to spawn approach.

I assume this is a simple problem being able to be solved here quick. This is a common problem. I've tested on branch master and this PR. (What about another issue?)

**Comment by hhyyrylainen** [comment]
Do you mean that, that screenshot is the situation on this PR's version? If so, then I guess this PR would be good if this addressed the too much stuff on screen at once problem.

**Comment by 84634E1A607A** [comment]
This screenshot is taken on branch master. But similar situation can happen on this PR. Here's the screenshot:
image

**Comment by hhyyrylainen** [comment]
The author of this PR said that they don't have time to finish this. So this PR is now looking for another person to finish it.

**Comment by hhyyrylainen** [comment]
A new PR is going to replace this: https://github.com/Revolutionary-Games/Thrive/pull/2397
